src/lib.py

```diff
@@ -496,47 +496,63 @@ def get_kit_coverage(dbo, pid):
 # does not consider the endpoints
 # ranges must be sorted on their minaddr
 # input vector must be sorted
-def in_range(v_vect, ranges):
+def in_range(v_vect, ranges, spans):
     c_vect = []
     ii = 0
     nmax = len(ranges)
-    for v in v_vect:
+    for iv,v in enumerate(v_vect):
         while ii < nmax and ranges[ii][0] < v:
             ii += 1
         if ii > 0 and v > ranges[ii-1][0] and v < ranges[ii-1][1]:
-            c_vect.append(True)
+            if spans and v+spans[iv] < ranges[ii-1][1]:
+                # span fits within the range
+                c_vect.append(True)
+            elif spans:
+                # span exceeded the upper end of the range
+                c_vect.append(False)
+            else:
+                # no span - SNP fits within range
+                c_vect.append(True)
         else:
             c_vect.append(False)
     if len(c_vect) != len(v_vect):
         raise ValueError
     return c_vect
 
-
-# check kit coverage for a vector of variants return a vector of BED range
-# coverage for a person that represents if the call is a) in a range, b) on the
-# lower edge of a range, c) on the upper edge of a range, or d) not covered by
-# a range
-def get_call_coverage(dbo, pid, vids):
+# Check kit coverage for a vector of variants; return a coverage vector that
+# indicates if that person has a BED range such that the call is a) in a range,
+# b) on the lower edge of a range, c) on the upper edge of a range, or d) not
+# covered by a range. If spans is passed, it corresponds to the maximum length
+# affected by the variant in vids: 1 for SNPs, max(ref,alt) for indels
+def get_call_coverage(dbo, pid, vids, spans=None):
     # FIXME currently only returns True/False, doesn't handle range ends
+    # FIXME maybe more efficient to pass positions instead of ids
     dc = dbo.cursor()
     dc.execute('drop table if exists tmpt')
     dc.execute('create table tmpt (vid integer)')
+    trace(3, 'get_call_coverage: vids:{}...'.format(vids[:10]))
+    if spans:
+        trace(3, 'get_call_coverage: spans:{}...'.format(spans[:10]))
     dc.executemany('insert into tmpt values(?)', [(a,) for a in vids])
     calls = dc.execute('''select v.id, v.pos from variants v
                           inner join tmpt t on t.vid=v.id
                           order by 2''')
+    # form a list of positions of interest
     cv = [v[1] for v in calls]
     trace(500, '{} calls: {}...'.format(len(cv), cv[:20]))
     rc = dbo.cursor()
     ranges = rc.execute('''select minaddr,maxaddr from bedranges r
                            inner join bed b on b.bID=r.id
                            where b.pid=?
                            order by 1''', (pid,))
+    # form a list of ranges of interest
     rv = [v for v in ranges]
     if len(rv) == 0:
         return []
     trace(500, '{} ranges: {}...'.format(len(rv), rv[:20]))
-    coverage = in_range(cv, rv)
+    coverage = in_range(cv, rv, spans)
+    dc.close()
```

Comment:
does a cursor auto-close when it is destroyed by going out of scope?
if it does, then explicitly closing it on exit is just clutter.

Reply:
sqlite3 calls in general need some cleaning up. It's true close probably should not be called here, though cursors I think do not get auto closed when they leave context unless the connection does. The whole area needs some review. Sqlite3 is currently embraced as a way to quickly prototype, but the code should probably be portable to some other db class, and I'm not sure that's currently the case. I will look at this closer in some future pull request.

Reply:
https://www.python.org/dev/peps/pep-0249/#cursor-objects says (or at least implies) that a cursor is closed when __del__ is called on it, which should be when it's destroyed. Seems OK to leave off closing when the cursor variable has local scope.